File: Python/tools/ofdm_wavgen.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 17 08:50:42 2022

OFDM waveform generator

@author: Ryan Tsai
"""

import math
import numpy as np
from scipy import signal
from scipy import fft

def ofdm_wavgen(nsym,bw,scs,num_sc,start_sc,modorder,en_tprecode,ncp=0,wola=0,seed=[]):
    if seed == []:
        rng = np.random.default_rng()
    else:
        rng = np.random.default_rng(seed)
    
    nrb = bw*5 # max number of resource blocks
    nfft = 2**math.ceil(math.log2(nrb*12))
    ncp = round(nfft*ncp/100)
    sym_len = nfft+ncp
    fs = nfft*scs/1000 # waveform sampling rate
    bps = round(math.log2(modorder))
    wola_len = round(nfft*wola/100)
    wola_len = wola_len + (wola_len % 2) # needs to be even (half outside of symbol, half inside of symbol)
    b = generate_wola_window(wola_len,sym_len)
    
    x_standard = np.zeros(nsym*sym_len) + 1j*np.zeros(nsym*sym_len) # no WOLA
    x = np.zeros(nsym*sym_len+wola_len) + 1j*np.zeros(nsym*sym_len+wola_len) # with WOLA
    for sdx in range(nsym):
        bitstream = rng.random(bps*num_sc).round() # bits for current symbol
        tones = modulation_mapper(bitstream,modorder) # convert from bits to modulated tones
        if en_tprecode: tones = fft.fft(tones) # transform precoding
        tones_nrb = np.zeros(nrb*12) + 1j*np.zeros(nrb*12) # available tones
        tones_nrb[start_sc:start_sc+num_sc] = tones # allocated tones
        tones_all = np.concatenate((np.zeros(round((nfft-nrb*12)/2)), tones_nrb, np.zeros(round((nfft-nrb*12)/2)))) # freq-domain zero-padding
        tones_all = np.concatenate((tones_all[round(nfft/2):], tones_all[0:round(nfft/2)])) # FFT shift prior to IFFT
        sym = fft.ifft(tones_all) # OFDM symbol
        sym_cp = np.concatenate((sym[nfft-ncp:], sym)) # add CP
        x_standard[sdx*sym_len:(sdx+1)*sym_len] = sym_cp # standard waveform
        
        # Apply WOLA
        sym_wola = np.concatenate((sym[round(nfft-ncp-wola_len/2):], sym, sym[0:round(wola_len/2)]))
        sym_wola = np.multiply(sym_wola,b)
        x[sdx*sym_len:(sdx+1)*sym_len+wola_len] = sym_wola + x[sdx*sym_len:(sdx+1)*sym_len+wola_len] # waveform with WOLA
        
        # Save allocated tone indices for demodulation
        if sdx == 0:
            tone_idx = np.ones(nrb*12) # 1 = unallocated tones (IBE)
            tone_idx[start_sc:start_sc+num_sc] = 2 # 2 = allocated tones (EVM, equalizer)
            tone_idx = np.concatenate((np.zeros(round((nfft-nrb*12)/2)), tone_idx, np.zeros(round((nfft-nrb*12)/2)))) # 0 = zero-padded tones
            tone_idx = np.concatenate((tone_idx[round(nfft/2):], tone_idx[0:round(nfft/2)]))
    
    # Output dictionary for demodulation
    cfg = {}
    cfg['fs'] = fs; cfg['nfft'] = nfft; cfg['ncp'] = ncp; cfg['nrb'] = nrb; cfg['nsym'] = nsym
    cfg['bw'] = bw; cfg['scs'] = scs; cfg['num_sc'] = num_sc; cfg['start_sc'] = start_sc
    cfg['modorder'] = modorder; cfg['en_tprecode'] = en_tprecode; cfg['tone_idx'] = tone_idx
    cfg['wola_len'] = wola_len

    return (x,x_standard,cfg)
            
def modulation_mapper(bitstream,modorder):
    if modorder == 4:
        tones = 1-2*bitstream[0::2] + 1j*(1-2*bitstream[1::2])
        tones = tones/math.sqrt(2)
    elif modorder == 16:
        tones = (1-2*bitstream[0::4]) * (2-(1-2*bitstream[2::4])) + 1j * (1-2*bitstream[1::4]) * (2-(1-2*bitstream[3::4]))
        tones = tones/math.sqrt(10)
    elif modorder == 64:
        tones = (1-2*bitstream[0::6]) * (4-(1-2*bitstream[2::6])*(2-(1-2*bitstream[4::6]))) + 1j * (1-2*bitstream[1::6]) * (4-(1-2*bitstream[3::6])*(2-(1-2*bitstream[5::6])))
        tones = tones/math.sqrt(42)
    elif modorder == 256:
        tones = (1-2*bitstream[0::8]) * (8-(1-2*bitstream[2::8])*(4-(1-2*bitstream[4::8])*(2-(1-2*bitstream[6::8])))) + 1j * (1-2*bitstream[1::8]) * (8-(1-2*bitstream[3::8])*(4-(1-2*bitstream[5::8])*(2-(1-2*bitstream[7::8]))))
        tones = tones/math.sqrt(170)
        
    return tones

def generate_wola_window(wola_len,sym_len):
    # sym_len = nfft+ncp
    
    if wola_len > 0:
        b = signal.hann(2*wola_len)
        b = np.concatenate((b[0:wola_len], np.ones(sym_len-wola_len), b[wola_len:]))
    else:
        b = np.ones(sym_len)
        
    return b
        

File: Octave/tools/ofdm_wavgen.m
% Author: Ryan Tsai
% Created: 2022-03-30

function [x,x_standard,cfg] = ofdm_wavgen(nsym,bw,scs,num_sc,start_sc,modorder,en_tprecode,ncp,wola,seed)
% Generates OFDM waveform with 1-subcarrier resolution in subcarrier allocation
% bw = BW in MHz
% scs = SCS in kHz
% ncp = CP length in percentage of nfft
% wola = WOLA length in percentage of nfft

if ~exist('ncp','var'), ncp = 0; endif
if ~exist('wola','var'), wola = 0; endif
if exist('seed','var'), rand("state",seed); endif

nrb = bw*5; % max number of resource blocks
nfft = 2^ceil(log2(nrb*12));
ncp = round(nfft*ncp/100);
sym_len = nfft+ncp;
fs = nfft*scs/1000; % waveform sampling rate
bps = log2(modorder); % bits per tone
wola_len = round(nfft*wola/100);
wola_len = wola_len+mod(wola_len,2);
b = generate_wola_window(wola_len,sym_len);

x_standard = zeros(1,nsym*sym_len); % no WOLA
x = zeros(1,nsym*sym_len+wola_len); % with WOLA
for sdx = 1:nsym
  bitstream = rand(1,bps*num_sc)-0.5>0;
  tones = modulation_mapper(bitstream,modorder);
  if en_tprecode, tones = fft(tones); endif
  tones_nrb = zeros(1,nrb*12);
  tones_nrb(1+start_sc:start_sc+num_sc) = tones;
  tones_all = [zeros(1,(nfft-nrb*12)/2) tones_nrb zeros(1,(nfft-nrb*12)/2)];
  tones_all = [tones_all(nfft/2+1:end) tones_all(1:nfft/2)];
  sym = ifft(tones_all);
  sym_cp = [sym(nfft-ncp+1:end) sym];
  x_standard(1+(sdx-1)*sym_len:sdx*sym_len) = sym_cp;
  
  sym_wola = [sym(nfft-ncp-wola_len/2+1:end) sym sym(1:wola_len/2)];
  sym_wola = sym_wola.*b;
  x(1+(sdx-1)*sym_len:sdx*sym_len+wola_len) = sym_wola + x(1+(sdx-1)*sym_len:sdx*sym_len+wola_len);
  
  if sdx == 1
    tone_idx = ones(1,nrb*12);
    tone_idx(1+start_sc:start_sc+num_sc) = 2;
    tone_idx = [zeros(1,(nfft-nrb*12)/2) tone_idx zeros(1,(nfft-nrb*12)/2)];
    tone_idx = [tone_idx(nfft/2+1:end) tone_idx(1:nfft/2)];
  endif
endfor

%x1 = [zeros(1,wola_len/2) x_standard zeros(1,wola_len/2)];
%len = 2*(wola_len+nfft);
%figure; hold on; plot(abs(x1(1:len))); plot(abs(x(1:len)));

% Struct for calculating EVM
cfg.fs = fs; cfg.nfft = nfft; cfg.ncp = ncp; cfg.nrb = nrb; cfg.nsym = nsym;
cfg.bw = bw; cfg.scs = scs; cfg.num_sc = num_sc; cfg.start_sc = start_sc;
cfg.modorder = modorder; cfg.en_tprecode = en_tprecode; cfg.tone_idx = tone_idx;
cfg.wola_len = wola_len;

endfunction

function tones = modulation_mapper(bitstream,modorder)
  if modorder == 4
    tones = 1-2*bitstream(1:2:end) + 1j*(1-2*bitstream(2:2:end));
    tones = tones/sqrt(2);
  endif
endfunction

function b = generate_wola_window(wola_len,sym_len)
% sym_len = nfft+ncp

if wola_len > 0
  b = hanning(2*wola_len).';
  b = [b(1:wola_len) ones(1,sym_len-wola_len) b(wola_len+1:end)];
else
  b = ones(1,sym_len);
endif
%b = hanning(wola_len+sym_len).';
  
endfunction
